implementation-details.md

```diff
@@ -22,16 +23,31 @@ Each Solana node participates in gossip in one of two modes:
+The node sends an IP echo server request with a list of UDP and TCP ports whose availability needs to be checked by the entrypoint node.
+The entrypoint node tries to establish a connection with the requesting node's TCP ports and sends packets to the UDP ports. The entrypoint node then replies back with the IP echo server response containing the public address of the requesting node and the cluster shred version.
+
+The node starts the Gossip service and then advertises itself to the cluster by sending messages to the peer identified by the entrypoint. The node sends:
+* two [push messages](#push-messages) containing its own [`Version`](gossip-protocol-spec.md#version) and [`NodeInstance`](gossip-protocol-spec.md#nodeinstance),
+* a [pull request](#pull-requests) that contains the node's own [`LegacyContactInfo`](gossip-protocol-spec.md#legacycontactinfo).
+
+The push messages are [propagated further](#propagating-push-messages-over-a-cluster) to other nodes in the cluster. Soon after that, the node should start receiving [pull responses](#pull-responses), push messages, and [pings](#ping-and-pong-messages) from other nodes.
 
 When the connection with the cluster is established, the node can fully participate in the gossip - it can update the cluster with its data (push messages) and receive data from its peers through received push messages and by sending pull requests and receiving the corresponding pull responses.
 
+However, for the node to switch from the default `spy` mode to the `gossip` mode, it awaits a `SnapshotHashes` message from the cluster.
+Once it receives the message, the node will try to download the snapshots from any known RPC servers in the cluster.
+After the snapshot downloads, the node runs in `gossip` mode.
+
 ### Keeping the connection alive
-To keep the connection with other nodes, nodes must constantly exchange messages. Nodes in the cluster send around 30-40 pull requests per second and push messages every 7.5 seconds. If a node stops sending pull requests and push messages for more than 15 seconds it will be ignored by other nodes and will not receive any push messages or pull requests. Sending ping messages only is not enough, node needs to send push messages or pull requests at least every 15 seconds to keep the connection alive.
+To keep the connection with other nodes, nodes must constantly exchange messages.
+Nodes in the cluster send lots of pull requests per second and push messages every 7.5 seconds.
+If a node stops sending pull requests and push messages for more than 15 seconds it will be ignored by other nodes and will not receive any push messages or pull requests.
+Sending ping messages only is not enough. The node needs to send pull requests at least every 15 seconds and push messages where some push messages must contain the node's `NodeInstance` and `ContactInfo` to keep the connection alive.
```

Comment:
> ...The node needs to send pull requests at least every 15 seconds and push messages where some push messages must contain the node's `NodeInstance` and `ContactInfo` to keep the connection alive.

...The node must send its `ContactInfo` at least every 15 seconds to keep the connection alive. The node can do this through PushMessages or a combination of PushMessages and PullRequests.

Just ran a test and a node doesn't have to send `NodeInstance` to stay connected. Just `ContactInfo`